// @jellyfish-dev/membrane-webrtc-js/dist/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultSimulcastBitrates = exports.defaultBitrates = exports.simulcastTransceiverConfig = void 0;
// const TEMPORAL_LAYERS_COUNT = 2;
exports.simulcastTransceiverConfig = {
    direction: "sendonly",
    // keep this array from low resolution to high resolution
    // in other case lower resolution encoding can get
    // higher max_bitrate
    sendEncodings: [
        {
            rid: "l",
            active: false,
            // maxBitrate: 4_000_000,
            scaleResolutionDownBy: 4.0,
            //   scalabilityMode: "L1T" + TEMPORAL_LAYERS_COUNT,
        },
        {
            rid: "m",
            active: false,
            scaleResolutionDownBy: 2.0,
        },
        {
            rid: "h",
            active: false,
            // maxBitrate: 4_000_000,
            // scalabilityMode: "L1T" + TEMPORAL_LAYERS_COUNT,
        },
    ],
};
exports.defaultBitrates = { audio: 50000, video: 2500000 };
exports.defaultSimulcastBitrates = {
    h: 2500000,
    m: 500000,
    l: 150000,
};
